qutip/nonmarkov/bofin.py

```diff
@@ -0,0 +1,1445 @@
+            )
+        else:
+            L = L.tocsc()
+            LU = splu(L)
+            solution = LU.solve(b_mat)
+
+        data = dense2D_to_fastcsr_fmode(vec2mat(solution[:n ** 2]), n, n)
+        data = 0.5 * (data + data.H)
+
+        solution = solution.reshape((self._n_ados, n, n))
+
+        return Qobj(data, dims=self._H0.dims), solution
+
+    def run(self, rho0, tlist, ado_init=False, ado_return=False):
+        """
+        Solve for the time evolution of the system.
+
+        Parameters
+        ----------
+        rho0 : Qobj or numpy.array
+            Initial state (density matrix) of the system
+            if ``full_init`` is ``False``.
+            If ``full_init`` is ``True``, then ``rho0`` should be a numpy array
+            of the initial state and the initial state of all ADOs. Usually
+            the state of the ADOs would be determine from a previous call
+            to ``.run(..., full_return=True)``.
+
+        tlist : list
+            An ordered list of times at which to return the value of the state.
+
+        ado_init: bool, default False
+            Indicates if initial condition is just the system state, or a
+            numpy array including all ADOs.
+
+        ado_return: bool, default True
+            Whether to also return as output the full state of all ADOs.
+
+        Returns
+        -------
+        :class:`qutip.solver.Result`
+            The results of the simulation run.
+            The times (tlist) are stored in ``result.times``.
+            The state at each time is stored in ``result.states``.
+            If ``ado_return`` is ``True``, then the full ADO state at each
+            time is stored in ``result.ado_states``.
+            The state of a particular ADO may be extracted from
+            ``result.ado_states[i]`` by calling :meth:`.extract_ado`.
+        """
+        n = self._sys_shape
+        rho_shape = (n, n)
+        rho_dims = self._sys_dims
+        hierarchy_shape = (self._n_ados, n, n)
+
+        output = Result()
+        output.solver = "HEOMSolver"
+        output.times = tlist
+        output.states = []
+
+        if ado_init:
+            rho0_he = rho0
+        else:
+            rho0_he = np.zeros([n ** 2 * self._n_ados], dtype=complex)
+            rho0_he[:n ** 2] = rho0.full().ravel('F')
+
+        if ado_return:
+            output.ado_states = []
+
+        solver = self._ode
+        solver.set_initial_value(rho0_he, tlist[0])
+
+        self.progress_bar.start(len(tlist))
+        for t_idx, t in enumerate(tlist):
+            self.progress_bar.update(t_idx)
+            if t_idx != 0:
+                solver.integrate(t)
+            rho = Qobj(
+                solver.y[:n ** 2].reshape(rho_shape, order='F'),
+                dims=rho_dims,
+            )
+            output.states.append(rho)
+            if ado_return:
+                output.ado_states.append(solver.y.reshape(hierarchy_shape))
```

Comment:
very minor comment:  this reshape is useful for extracting info from the ados, but i think is incompatible with then using this output as input for ado_inti in .run()?